cmd/syncthing/connections.go

```diff
@@ -266,45 +255,103 @@ func (s *connectionSvc) connect() {
 			}
 
 			for _, addr := range addrs {
-				host, port, err := net.SplitHostPort(addr)
-				if err != nil && strings.HasPrefix(err.Error(), "missing port") {
-					// addr is on the form "1.2.3.4"
-					addr = net.JoinHostPort(addr, "22000")
-				} else if err == nil && port == "" {
-					// addr is on the form "1.2.3.4:"
-					addr = net.JoinHostPort(host, "22000")
+				uri, err := url.Parse(addr)
+				if err != nil {
+					l.Infoln("Failed to parse connection url:", addr, err)
+					continue
 				}
+
+				dialer, ok := dialers[uri.Scheme]
+				if !ok {
+					l.Infoln("Unknown address schema", uri.String())
+					continue
+				}
+
 				if debugNet {
-					l.Debugln("dial", deviceCfg.DeviceID, addr)
+					l.Debugln("dial", deviceCfg.DeviceID, uri.String())
+				}
+				conn, err := dialer(uri, s.tlsCfg)
+				if err != nil {
+					if debugNet {
+						l.Debugln("dial failed", deviceCfg.DeviceID, uri.String(), err)
+					}
+					continue
+				}
+
+				if connected {
+					s.model.Close(deviceID, fmt.Errorf("switching connections"))
 				}
 
-				raddr, err := net.ResolveTCPAddr("tcp", addr)
+				s.conns <- intermediateConnection{
+					conn, model.ConnectionTypeBasicDial,
+				}
+				continue nextDevice
+			}
+
+			// Only connect via relays if not already connected
+			if connected {
+				continue nextDevice
+			}
+
+			last, ok := s.lastRelayCheck[deviceID]
+			if ok && last.Add(time.Duration(s.cfg.Options().RelayReconnectIntervalM)*time.Minute).After(time.Now()) {
+				if debugNet {
+					l.Debugln("Skipping connecting via relay to", deviceID, "last checked at", last)
+				}
+				continue nextDevice
+			} else if debugNet {
+				l.Debugln("Trying relay connections to", deviceID, relays)
+			}
+
+			s.lastRelayCheck[deviceID] = time.Now()
+
+			for _, addr := range relays {
+				uri, err := url.Parse(addr)
+				if err != nil {
+					l.Infoln("Failed to parse relay connection url:", addr, err)
+					continue
+				}
+
+				inv, err := client.GetInvitationFromRelay(uri, deviceID, s.tlsCfg.Certificates)
 				if err != nil {
 					if debugNet {
-						l.Debugln(err)
+						l.Debugf("Failed to get invitation for %s from %s: %v", deviceID, uri, err)
 					}
 					continue
+				} else if debugNet {
+					l.Debugln("Succesfully retrieved relay invitation for", deviceID, "from", uri)
 				}
 
-				conn, err := net.DialTCP("tcp", nil, raddr)
+				to := fmt.Sprintf("%s:%d", net.IP(inv.Address).String(), inv.Port)
```

Comment:
This would be nice as `SessionInvitation.AddressString()` or so, or just a `SessionInvitation.String()` that makes sense so we could just drop `%v` for the `inv` in the debug prints below..